nntrainer/dataset/batch_queue.h

```diff
@@ -91,11 +96,184 @@ class BatchQueue {
   std::condition_variable_any q_reader_cv;
   std::condition_variable_any q_writer_cv;
 
+  std::queue<std::unique_ptr<T>> q;
+};
+
+/**
+ * @brief A view container that calls a callback on destruct
+ * @note the callback must be noexcept, and the given underlying data must
+ * outleave the lifetime of this class
+ *
+ * @tparam T underlying type
+ */
+template <typename T> class ScopedView {
+public:
   /**
-   * @todo consider using circular buffer if this is too slow
+   * @brief Construct a new Scoped View object
    *
+   * @param data_ reference of the underlying data
+   * @param on_notify_ callback to be called on exit, this is not copied but
+   * reused
    */
-  std::queue<std::unique_ptr<T>> q;
+  ScopedView(T *data_, std::function<void(void)> &&on_notify_) :
+    data(data_),
+    on_notify(std::forward<std::function<void(void)>>(on_notify_)) {}
+
+  ScopedView(const ScopedView &rhs) = delete;
```

Comment:
i wonder why delete this?

Reply:
ScopedView calls a function on destroy, this has to be done once, so need unique ownership to the view.